### PR TITLE
[SPARK-27119][SQL] Do not infer schema when reading Hive serde table with native data source

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

In Spark 2.1, we hit a correctness bug. When reading a Hive serde parquet table with the native parquet data source, and the actual file schema doesn't match the table schema in Hive metastore(only upper/lower case difference), the query returns 0 results.

The reason is that, the parquet reader is case sensitive. If we push down filters with column names that don't match the file physical schema case-sensitively, no data will be returned.

To fix this bug, there were 2 solutions proposed at that time:
1. Add a config to optionally disable parquet filter pushdown, and make parquet column pruning case insensitive.
https://github.com/apache/spark/pull/16797

2. Infer the actual schema from data files, when reading Hive serde table with native data source. A config is provided to disable it.
https://github.com/apache/spark/pull/17229

Solution 2 was accepted and merged to Spark 2.1.1

In Spark 2.4, we refactored the parquet data source a little:
1. do parquet filter pushdown with the actual file schema.
https://github.com/apache/spark/pull/21696

2. make parquet filter pushdown case insensitive.
https://github.com/apache/spark/pull/22197

3. make parquet column pruning case insensitive.
https://github.com/apache/spark/pull/22148

With these patches, the correctness bug in Spark 2.1 no longer exists, and the schema inference becomes unnecessary.

To be safe, this PR just changes the default value to NEVER_INFER, so that users can set it back to INFER_AND_SAVE. If we don't receive any bug reports for it, we can remove the related code in the next release.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @budde @wangyum @xuanyuanking

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -596,7 +597,7 @@ object SQLConf {
     .stringConf
     .transform(_.toUpperCase(Locale.ROOT))
     .checkValues(HiveCaseSensitiveInferenceMode.values.map(_.toString))
-    .createWithDefault(HiveCaseSensitiveInferenceMode.INFER_AND_SAVE.toString)
+    .createWithDefault(HiveCaseSensitiveInferenceMode.NEVER_INFER.toString)
```
Comment (xuanyuanking) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: https://github.com/apache/spark/pull/24041/files#diff-9a6b543db706f1a90f790783d6930a13R592 nit for the doc, the INFER_AND_SAVE is no longer the default. Should we mention this in migrate guide?

**Review by xuanyuanking — APPROVED** [review]
+1 for me. Thanks for the nice PR description, very clear.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: +1 for the comments; 1) update the config description at line 592 ~ 593 and 2) `sql-migration-guide-upgrade.md`

File: docs/sql-migration-guide-upgrade.md [hunk 2]
```diff
@@ -89,18 +89,20 @@ displayTitle: Spark SQL Upgrading Guide
...
 
-    - CSV/JSON datasources use java.time API for parsing and generating CSV/JSON content. In Spark version 2.4 and earlier, java.text.SimpleDateFormat is used for the same purpose with fallbacks to the parsing mechanisms of Spark 2.0 and 1.x. For example, `2018-12-08 10:39:21.123` with the pattern `yyyy-MM-dd'T'HH:mm:ss.SSS` cannot be parsed since Spark 3.0 because the timestamp does not match to the pattern but it can be parsed by earlier Spark versions due to a fallback to `Timestamp.valueOf`. To parse the same timestamp since Spark 3.0, the pattern should be `yyyy-MM-dd HH:mm:ss.SSS`.
+  - CSV/JSON datasources use java.time API for parsing and generating CSV/JSON content. In Spark version 2.4 and earlier, java.text.SimpleDateFormat is used for the same purpose with fallbacks to the parsing mechanisms of Spark 2.0 and 1.x. For example, `2018-12-08 10:39:21.123` with the pattern `yyyy-MM-dd'T'HH:mm:ss.SSS` cannot be parsed since Spark 3.0 because the timestamp does not match to the pattern but it can be parsed by earlier Spark versions due to a fallback to `Timestamp.valueOf`. To parse the same timestamp since Spark 3.0, the pattern should be `yyyy-MM-dd HH:mm:ss.SSS`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide-upgrade.md]: just fix the indentation

**Comment by wangyum** [comment]
retest this please

File: docs/sql-migration-guide-upgrade.md (hunk 2 above)
Comment (HyukjinKwon) [reply docs/sql-migration-guide-upgrade.md]: @cloud-fan, I think this indentation was intentional to make the notes collapsed to group Proleptic Gregorian calendar changes.

**Comment by HyukjinKwon** [comment]
Yea, very nice PR description.

**Review by HyukjinKwon — APPROVED** [review]

File: docs/sql-migration-guide-upgrade.md (hunk 2 above)
Comment (cloud-fan) [reply docs/sql-migration-guide-upgrade.md]: damn my editor didn't detect this.... let me revert it

**Review by gatorsmile — COMMENTED** [review]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**